File: yandex_eserajim_sleepless_nights/2. Классы подобия треугольников/solution_.py
import sys
from typing import List, Set

def gcd(a: int, b: int) -> int:
    while b:
        a, b = b, a % b
    return a

def canonical_representation(a: int, b: int, c: int) -> str:
    sides: List = sorted([a, b, c])
    common_gcd = gcd(gcd(sides[0], sides[1]), sides[2])
    return f"{sides[0] // common_gcd}-{sides[1] // common_gcd}-{sides[2] // common_gcd}"

def main() -> None:
    n = int(input())
    classes: Set[str] = set()

    for _ in range(n):
        parts = input().split()
        a, b, c = int(parts[0]), int(parts[1]), int(parts[2])
        repr_str = canonical_representation(a, b, c)
        classes.add(repr_str)

    print(len(classes))

if __name__ == "__main__":
    main()

File: yandex_eserajim_sleepless_nights/2. Классы подобия треугольников/solution.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program 
{
    static long GCD(long a, long b)
    {
        while (b != 0)
        {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    static string CanonicalRepresentation(long a, long b, long c)
    {
        long[] sides = { a, b, c };
        Array.Sort(sides);
        long commonGCD = GCD(GCD(sides[0], sides[1]), sides[2]);
        return $"{sides[0]/commonGCD}-{sides[1]/commonGCD}-{sides[2]/commonGCD}";
    }

    static void Main()
    {
        int n = int.Parse(Console.ReadLine());
        HashSet<string> classes = new HashSet<string>();

        for (int i = 0; i < n; i++)
        {   
            string[] parts = Console.ReadLine().Split();
            long a = long.Parse(parts[0]);
            long b = long.Parse(parts[1]);
            long c = long.Parse(parts[2]);
            
            string repr = CanonicalRepresentation(a, b, c);
            classes.Add(repr);
        }
        Console.WriteLine(classes.Count);
    }
}